source/extensions/common/aws/credentials_provider_impl.cc

```diff
@@ -62,13 +72,85 @@ Credentials EnvironmentCredentialsProvider::getCredentials() {
   return Credentials(access_key_id, secret_access_key, session_token);
 }
 
-void MetadataCredentialsProviderBase::refreshIfNeeded() {
+void CachedCredentialsProviderBase::refreshIfNeeded() {
   const Thread::LockGuard lock(lock_);
   if (needsRefresh()) {
     refresh();
   }
 }
 
+bool CredentialsFileCredentialsProvider::needsRefresh() {
+  return api_.timeSource().systemTime() - last_updated_ > REFRESH_INTERVAL;
+}
+
+std::string GetEnvironmentVariableOrDefault(const std::string& variable_name,
+                                            const std::string& default_value) {
+  const char* value = getenv(variable_name.c_str());
+  return (value != NULL) && (value[0] != '\0') ? value : default_value;
+}
+
+void CredentialsFileCredentialsProvider::refresh() {
+  ENVOY_LOG(debug, "Getting AWS credentials from the credentials file");
+
+  const auto credentials_file = GetEnvironmentVariableOrDefault(
+      AWS_SHARED_CREDENTIALS_FILE, DEFAULT_AWS_SHARED_CREDENTIALS_FILE);
+  const auto profile = GetEnvironmentVariableOrDefault(AWS_PROFILE, DEFAULT_AWS_PROFILE);
+
+  extractCredentials(credentials_file, profile);
+}
+
+void CredentialsFileCredentialsProvider::extractCredentials(const std::string& credentials_file,
+                                                            const std::string& profile) {
+  if (credentials_file.empty() || profile.empty()) {
+    return;
+  }
+
+  std::ifstream file(credentials_file);
+  if (!file) {
+    ENVOY_LOG(error, "Error opening credentials file {}", credentials_file);
+    return;
+  }
+
+  std::string access_key_id, secret_access_key, session_token;
+  std::string profile_start = fmt::format("[{}]", profile);
+
+  bool found_profile = false;
+  std::string line;
+  while (std::getline(file, line)) {
```

Comment:
can we trim the line for whitespace and new line character?
`continue` if line is empty.

Reply:
Done.